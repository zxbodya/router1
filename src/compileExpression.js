//todo: tests for expression language
const compileExpression = (expression)=> {
  //todo: for non-production verify expression syntax
  //todo: if syntax is not valid show messages about typical errors

  // http://stackoverflow.com/questions/8844256/split-string-including-regular-expression-match
  let textParts = expression.split(/<([^>]+?)(?::([^>]+))?>/);
  let matcher = [];
  let generateParts = [];
  let paramNames = [];
  for (let i = 0, l = textParts.length; i < l; i++) {
    let part = textParts[i];
    if (i % 3 === 0) {
      //  http://stackoverflow.com/questions/3446170/escape-string-for-use-in-javascript-regex#answer-6969486
      matcher.push(part.replace(/[\-\[\]\/\{\}\(\)\*\+\?\.\\\^\$\|]/g, "\\$&"));
      generateParts.push(part);
    }
    if (i % 3 === 1) {
      paramNames.push(part);
      generateParts.push(null);
    }
    if (i % 3 === 2) {
      //todo: verify expression if it has captures - throw an error
      matcher.push('(', part ? part : '.*', ')');
    }
  }
  return [matcher.join(''), generateParts, paramNames]
};

module.exports = compileExpression;
