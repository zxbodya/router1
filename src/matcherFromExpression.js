let matcherFromExpression = (info)=> {
  let [matcher,,paramNames]  = info;
  let matcherExp = new RegExp('^' + matcher + '$');

  return (path)=> {
    let matches, res = false;
    if (matches = path.match(matcherExp)) {
      //todo: verify matches count
      res = {};
      for (let i = 0, l = matches.length; i < l; i++) {
        res[paramNames[i]] = matches[i];
      }
    }
    return res;
  }
};

module.exports = matcherFromExpression;
