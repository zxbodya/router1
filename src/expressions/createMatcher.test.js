import compileExpression from './compile.js';
import matcherFromExpression from './createMatcher.js';

describe('Router, matcher from expression', function () {
  it('it matches strings without params', ()=> {
    let expression = compileExpression('aaaaa');
    let matcher = matcherFromExpression(expression);

    expect(matcher('aaaaa')).toEqual({});
    expect(matcher('aaaaaa')).toEqual(false);
    expect(matcher('/aaaaa')).toEqual(false);
    expect(matcher('aaaaa/')).toEqual(false);
  });

  it('it matches strings with one param', ()=> {
    let expression = compileExpression('aaaaa<a>');
    let matcher = matcherFromExpression(expression);

    expect(matcher('aaaaa123')).toEqual({a: '123'});
    expect(matcher('aaaaa321')).toEqual({a: '321'});
    expect(matcher('aaaaaa')).toEqual({a: 'a'});
    expect(matcher('1aaaaaa')).toEqual(false);
  });

  it('it generates strings with few params', ()=> {
    let expression = compileExpression('aaaaa<a>-<b>');
    let matcher = matcherFromExpression(expression);

    expect(matcher('aaaaa321-1')).toEqual({a: '321', b: '1'});
    expect(matcher('aaaaa321-444')).toEqual({a: '321', b: '444'});
    expect(matcher('aaaaa321-')).toEqual({a: '321', b: ''});
    expect(matcher('aaaaa-444')).toEqual({a: '', b: '444'});
    expect(matcher('aaaaa-')).toEqual({a: '', b: ''});
    expect(matcher('aaaaa')).toEqual(false);
  });

  it('it matches strings with param regexp', ()=> {
    let expression = compileExpression('aaaaa<a:\\d+>');
    let matcher = matcherFromExpression(expression);

    expect(matcher('aaaaa123')).toEqual({a: '123'});
    expect(matcher('aaaaa321')).toEqual({a: '321'});
    expect(matcher('aaaaaa')).toEqual(false);
    expect(matcher('aaaaa123a')).toEqual(false);
  });
});
