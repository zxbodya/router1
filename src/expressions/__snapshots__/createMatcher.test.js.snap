// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Router, matcher from expression decodes path before matching 1`] = `
Object {
  "a": "привіт",
  "b": "?#aZ09.-_~!$&\\\\'()*+,;=:@",
}
`;
